File: src/main/java/com/vandendaelen/nicephore/utils/OperatingSystems.java
package com.vandendaelen.nicephore.utils;

import com.vandendaelen.nicephore.clipboard.ClipboardManager;
import com.vandendaelen.nicephore.clipboard.impl.WindowsClipboardManagerImpl;

public enum OperatingSystems {
    WINDOWS(WindowsClipboardManagerImpl.getInstance()),
    LINUX(null),
    MAC(null),
    SOLARIS(null);

    private ClipboardManager manager;
    private static OperatingSystems os;

    OperatingSystems(ClipboardManager instance) {
        this.manager = instance;
    }

    public static OperatingSystems getOS() {
        if (os == null) {
            final String operSys = System.getProperty("os.name").toLowerCase();
            if (operSys.contains("win")) {
                os = OperatingSystems.WINDOWS;
            } else if (operSys.contains("nix") || operSys.contains("nux")
                    || operSys.contains("aix")) {
                os = OperatingSystems.LINUX;
            } else if (operSys.contains("mac")) {
                os = OperatingSystems.MAC;
            } else if (operSys.contains("sunos")) {
                os = OperatingSystems.SOLARIS;
            }
        }
        return os;
    }

    public ClipboardManager getManager() {
        return manager;
    }
}

File: common/src/main/kotlin/com/vandendaelen/nicephore/enums/OperatingSystems.kt
package com.vandendaelen.nicephore.enums

import com.vandendaelen.nicephore.clipboard.ClipboardManager
import com.vandendaelen.nicephore.clipboard.impl.MacOSClipboardManagerImpl
import com.vandendaelen.nicephore.clipboard.impl.WindowsClipboardManagerImpl

enum class OperatingSystems(val manager: ClipboardManager?) {
    WINDOWS(WindowsClipboardManagerImpl.instance),
    LINUX(null),
    MAC(MacOSClipboardManagerImpl.instance),
    SOLARIS(null);

    companion object {
        fun getOS(): OperatingSystems {
            val osName = System.getProperty("os.name", "").lowercase()
            return when {
                osName.contains("win") -> WINDOWS
                osName.contains("mac") || osName.contains("darwin") -> MAC
                osName.contains("sunos") || osName.contains("solaris") -> SOLARIS
                else -> LINUX
            }
        }
    }
}
